game/world/objects/npc.cpp

```diff
@@ -3479,10 +3500,17 @@ bool Npc::tickCast(uint64_t dt) {
   return true;
   }
 
-void Npc::endCastSpell() {
+void Npc::endCastSpell(bool abort) {
   int32_t castLvl = int(castLevel)-int(CS_Invest_0);
   if(castLvl<0)
     return;
+  if(abort) {
+    const SpellCode code = SpellCode(owner.script().invokeManaRelease(*this,currentTarget,manaInvested));
```

Comment:
>Apparently invokeManaRelease should be called when player releases action-button, and current logic is out of place.

It also needs to be accessed if mana reaches 0 in G1 because there's no mana check in script if total mana is enough to cast spell. I look for a cleaner solution.